File: functions/src/engine/assignment.ts
/**
 * Shared assignment helpers used by the reassign trigger and the assignment
 * callables: validate a specific member, award/move points, and notify an
 * assignee (calendar + email, stubbed until their phases).
 */

import { FieldValue } from "firebase-admin/firestore";
import { db } from "../lib/setup";
import { calendarService } from "../services/calendar";
import { emailService } from "../services/email";
import { appendActivity } from "../lib/log";
import { isBaseEligible } from "./assign";
import { RequestDoc, Settings, TeamMember } from "../types";

/** Request states in which assigned work counts as confirmed (points awarded). */
export const CONFIRMED_STATES = [
  "Request Accepted",
  "Event Covered",
  "Ready To post",
  "Posted",
];

export interface ValidateResult {
  ok: boolean;
  reason: string;
  member: TeamMember | null;
}

/** Validate a chosen member for a task (skill, active, strikes, campus, calendar). */
export async function validateMember(
  email: string,
  requiredSkill: string,
  atEvent: boolean,
  request: RequestDoc,
  settings: Settings
): Promise<ValidateResult> {
  const e = (email || "").toLowerCase();
  const snap = await db.doc(`team/${e}`).get();
  if (!snap.exists) {
    return { ok: false, reason: `${e} is not on the team`, member: null };
  }
  const member = { id: snap.id, ...(snap.data() as TeamMember) } as TeamMember;

  if (!isBaseEligible(member, requiredSkill, request, settings)) {
    return {
      ok: false,
      reason: `${member.name} is not eligible (inactive, at strike limit, lacks "${requiredSkill}", or different campus)`,
      member: null,
    };
  }
  if (atEvent && request.eventStart && request.eventEnd) {
    const free = await calendarService.isFree(
      e,
      request.eventStart,
      request.eventEnd
    );
    if (!free) {
      return {
        ok: false,
        reason: `${member.name} is busy during the event window`,
        member: null,
      };
    }
  }
  return { ok: true, reason: "", member };
}

/** Increment a member's points (negative to remove). Adds to an existing batch. */
export function awardPoints(
  batch: FirebaseFirestore.WriteBatch,
  email: string,
  delta: number
): void {
  if (!email || !delta) return;
  batch.update(db.collection("team").doc(email.toLowerCase()), {
    points: FieldValue.increment(delta),
  });
}

/** Calendar hold/reminder + email for a newly confirmed assignee (stubbed). */
export async function notifyAssignee(task: FirebaseFirestore.DocumentData): Promise<void> {
  if (task.atEvent && task.eventStart && task.eventEnd) {
    await calendarService.createHold({
      email: task.email,
      title: `${task.refCode} ${task.task} — ${task.eventName}`,
      start: task.eventStart,
      end: task.eventEnd,
      description: task.venue || "",
    });
  } else if (task.deadline) {
    await calendarService.createReminder({
      email: task.email,
      title: `${task.refCode} ${task.task} due — ${task.eventName}`,
      due: task.deadline,
    });
  }
  await emailService.send({
    to: task.email,
    subject: `[Assigned] ${task.refCode} ${task.task} — ${task.eventName}`,
    text: `You've been assigned as ${task.task} for ${task.eventName} (${task.refCode}).`,
  });
}

/**
 * Swap (or fill) a task's assignee to `newMember` and apply the side effects:
 * status, point move/award, coordinatorEmail propagation, notifications, log.
 * Used by both the reassign trigger and the assignment callables.
 */
export async function performSwap(
  taskRef: FirebaseFirestore.DocumentReference,
  task: FirebaseFirestore.DocumentData,
  newMember: TeamMember,
  request: RequestDoc
): Promise<void> {
  const oldEmail = (task.email || "").toLowerCase();
  const confirmedState = CONFIRMED_STATES.includes(request.status);
  const newStatus = confirmedState ? "CONFIRMED" : "PROPOSED";
  const points = task.points || 0;

  const batch = db.batch();
  batch.update(taskRef, {
    member: newMember.name,
    email: newMember.email,
    phone: newMember.phone || "",
    status: newStatus,
    pointsAwarded: confirmedState,
    reassignTo: "",
  });
  if (task.pointsAwarded && oldEmail) awardPoints(batch, oldEmail, -points);
  if (confirmedState) awardPoints(batch, newMember.email, points);

  if (task.task === "Event Coordinator") {
    batch.update(db.collection("requests").doc(task.requestId), {
      coordinatorEmail: newMember.email,
    });
    const siblings = await db
      .collection("tasks")
      .where("requestId", "==", task.requestId)
      .get();
    for (const d of siblings.docs) {
      if (d.id !== taskRef.id) {
        batch.update(d.ref, { coordinatorEmail: newMember.email });
      }
    }
  }
  await batch.commit();

  await notifyAssignee({
    ...task,
    member: newMember.name,
    email: newMember.email,
    phone: newMember.phone || "",
  });
  if (oldEmail && oldEmail !== newMember.email.toLowerCase()) {
    await emailService.send({
      to: oldEmail,
      subject: `[Reassigned] ${task.refCode} ${task.task}`,
      text: `Your ${task.task} task on ${task.refCode} (${task.eventName}) has been reassigned to ${newMember.name}.`,
    });
  }
  await appendActivity({
    event: "reassigned",
    requestId: task.requestId,
    refCode: task.refCode,
    member: newMember.email,
    detail: `${task.task}: ${oldEmail || "unfilled"} -> ${newMember.email}`,
  });
}

File: portal/engine/assignment.py
"""
Manual assignment and reassignment (docs/PRD.md §5.5).

Ported from `server/src/engine/assignment.ts`: validate a specific member for a
task, and swap a task's assignee with all the bookkeeping that implies — points
moved off the old holder and onto the new one, coordinator propagation,
notifications both ways, and an audit entry.
"""

from __future__ import annotations

from dataclasses import dataclass

from django.db import transaction

from core.activity import log_activity
from core.constants import RequestStatus, TaskStatus
from core.models import TeamMember
from services import email as email_service
from services.calendar import calendar_service

from .assign import is_base_eligible
from .notify import award_points, notify_assignee


@dataclass(frozen=True)
class Validation:
    ok: bool
    reason: str = ""
    member: TeamMember | None = None


def validate_member(
    email: str,
    required_skill: str,
    at_event: bool,
    request_obj,
    settings,
    *,
    require_skill: bool = True,
) -> Validation:
    """
    Check one hand-picked member against the same rules auto-assignment uses.

    `require_skill=False` is for a deliberate manual reassignment across
    verticals — active/strikes/campus/calendar checks still apply, only the
    skill match is dropped.
    """
    address = (email or "").strip().lower()
    member = TeamMember.objects.filter(email=address).first()
    if member is None:
        return Validation(ok=False, reason=f"{address} is not on the team")

    if not is_base_eligible(member, required_skill, request_obj, settings, require_skill=require_skill):
        skill_clause = f', lacking "{required_skill}"' if require_skill else ""
        return Validation(
            ok=False,
            reason=(
                f"{member.name} is not eligible — inactive, out of work, at the strike "
                f"limit{skill_clause}, or on a different campus"
            ),
        )

    if at_event and request_obj.event_start and request_obj.event_end:
        free = calendar_service().is_free(address, request_obj.event_start, request_obj.event_end)
        if not free:
            return Validation(ok=False, reason=f"{member.name} is busy during the event window")

    return Validation(ok=True, member=member)


def perform_swap(task, new_member, request_obj) -> None:
    """
    Move `task` to `new_member`, applying every consequence.

    Fills an UNFILLED task just as well as it replaces an assigned one — the
    only difference is whether there are points to claw back.
    """
    old_email = (task.email or "").lower()
    _commit_swap(task, new_member, request_obj, old_email)

    notify_assignee(task)
    if old_email and old_email != new_member.email.lower():
        email_service.send(
            old_email,
            f"[Reassigned] {task.ref_code} {task.task}",
            f"Your {task.task} task on {task.ref_code} ({task.event_name}) "
            f"has been reassigned to {new_member.name}.",
        )

    log_activity(
        "reassigned",
        request_obj=request_obj,
        ref_code=task.ref_code,
        member=new_member.email,
        detail=f"{task.task}: {old_email or 'unfilled'} -> {new_member.email}",
    )


@transaction.atomic
def _commit_swap(task, new_member, request_obj, old_email: str) -> None:
    from core.constants import TASK_EVENT_COORDINATOR
    from core.models import Task

    confirmed_state = request_obj.status in RequestStatus.CONFIRMED_STATES
    points = task.points or 0
    had_points = task.points_awarded

    task.member = new_member.name
    task.email = new_member.email
    task.phone = new_member.phone or ""
    task.status = TaskStatus.CONFIRMED if confirmed_state else TaskStatus.PROPOSED
    task.points_awarded = confirmed_state
    task.save(update_fields=["member", "email", "phone", "status", "points_awarded"])

    # The outgoing holder only loses points they were actually credited.
    if had_points and old_email:
        award_points(old_email, -points)
    if confirmed_state:
        award_points(new_member.email, points)

    # Replacing the coordinator re-points the whole request at the new one, or
    # every other task on it would still escalate to the person who left.
    if task.task == TASK_EVENT_COORDINATOR:
        request_obj.coordinator_email = new_member.email
        request_obj.save(update_fields=["coordinator_email"])
        Task.objects.filter(request=request_obj).exclude(pk=task.pk).update(
            coordinator_email=new_member.email
        )
